src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -3074,6 +2928,715 @@ void QgsOgrProvider::forceReload()
+            {
+              // Note: with KML is is possible that there are different Geometries types in the layer
+              geomType = layer_feature_type;
+              break;
+            }
+          }
+        }
+        if ( ( geomType == wkbUnknown ) && ( geomType_backup != wkbUnknown ) )
+        {
+          // Use this backup value in case we cannot find the layer-name
+          geomType = geomType_backup;
+        }
+      }
+    }
+  }
+  return geomType;
+}
+
+OGRwkbGeometryType QgsOgrProviderUtils::wkbSingleFlattenWrapper( OGRwkbGeometryType type )
+{
+  type = QgsOgrProviderUtils::wkbFlattenWrapper( type );
+  switch ( type )
+  {
+    case wkbMultiPoint:
+      return wkbPoint;
+    case wkbMultiLineString:
+      return wkbLineString;
+    case wkbMultiPolygon:
+      return wkbPolygon;
+    // no version-runtime checking needed here
+    case wkbMultiCurve:
+      return wkbCompoundCurve;
+    case wkbMultiSurface:
+      return wkbCurvePolygon;
+    default:
+      return type;
+  }
+}
+OGRwkbGeometryType QgsOgrProviderUtils::wkbFlattenWrapper( OGRwkbGeometryType eType )
+{
+  if ( wkbFlatten( eType ) == wkbUnknown )
+    return wkbUnknown;
+  return eType;
+}
+
+int QgsOgrProviderUtils::wkbHasZWrapper( OGRwkbGeometryType eType )
+{
+  return wkbHasZ( eType );
+}
+
+int QgsOgrProviderUtils::wkbHasMWrapper( OGRwkbGeometryType eType )
+{
+#if defined(GDAL_COMPUTE_VERSION) && GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,1,0)
+  return wkbHasM( eType );
+#else
+  if ( eType >= 2000 && eType < 3000 ) /* ISO M */
+    return 1;
+  if ( eType >= 3000 && eType < 4000 ) /* ISO ZM */
+    return 1;
+#endif
+  return 0;
+}
+
+OGRLayerH QgsOgrProvider::OGRGetLayerWrapper( OGRDataSourceH ogrDataSource, QString sLayerName, QString sGeometryName, long lLayerIndex )
+{
+  OGRLayerH ogrLayer = nullptr;
+  switch ( mOgrGetType )
+  {
+    case 0:
+    case 2:
+      // default: attempt to open with Layer-Name or with Geometry-name
+      mSubLayerString = QString( "%1:%2:%3:%4:%5:%6:%7" ).arg( mLayerIndex ).arg( mLayerName ).arg( mFeaturesCounted ).arg( QgsOgrProviderUtils::wkbGeometryTypeName( mOgrGeometryTypeFilter ) ).arg( sGeometryName ).arg( mGeometryIndex ).arg( mOgrGetType );
+      break;
+    case 1:
+      // non-unique-names need to be opened with the index value
+      mSubLayerString = QString( "%1:%2:%3:%4:%5:%6:%7" ).arg( mLayerIndex ).arg( mLayerName ).arg( mFeaturesCounted ).arg( QgsOgrProviderUtils::wkbGeometryTypeName( mOgrGeometryTypeFilter ) ).arg( sGeometryName ).arg( mGeometryIndex ).arg( mOgrGetType );
+      break;
+    default:
+      if ( mLayerIndex >= 0 )
+      {
+        if ( mDriverName != "SQLite" )
+        {
+          if ( ( !mLayerName.contains( "(" ) ) && ( !mLayerName.endsWith( QString( ")" ) ) ) )
+          {
+            if ( ( mGeometryName.isNull() ) || ( mGeometryName.isEmpty() ) )
+            {
+              mOgrGetType = 1; // read by layer-index [KML, GML with one geometry]
```

Comment:
Since the enum would describe the Functions being used in **QgsOgrProviderUtils::OgrGetType**
* it should be defined in QgsOgrProviderUtils

For this to avoid 'does not name a type' errors
* the definition of QgsOgrProviderUtils must be placed in and extra include file:
   * qgsogrproviderutils.h

which has now been done and the numeric values changed to the enums.